midend/unrollLoops.cpp

```diff
@@ -165,36 +165,52 @@ static long evalLoop(const IR::Expression *exp, cstring index, long val, bool &f
     return 0;
 }
 
+static const IR::PathExpression *incrRecursiveUse(const IR::Expression *e, cstring index) {
+    if (auto *pe = e->to<IR::PathExpression>()) {
+        if (pe->path->name == index) return pe;
+    } else if (auto *bin = e->to<IR::Operation::Binary>()) {
+        if (auto *pe = incrRecursiveUse(bin->left, index)) return pe;
+        if (auto *pe = incrRecursiveUse(bin->right, index)) return pe;
+    } else if (auto *un = e->to<IR::Operation::Unary>()) {
+        if (auto *pe = incrRecursiveUse(un->expr, index)) return pe;
+    }
+    return nullptr;
+}
+
+static bool find_def(const IR::IndexedVector<IR::StatOrDecl> &stmts, const IR::Node *def,
```

Comment:
Looks like it could be simplified by returning `a->to<IR::AssignmentStatement>()`, also the function itself could be renamed according to the code style.

Reply:
The reason the signature is the way it is (returning bool, with the actual node returned by pointer) was to make it easy to write the looped call at lines 199-202 -- when the return value is false, we don't want to overwrite the pointer that might have been returned by a previous call.   That in turn lets to two loops (199-202 and 209-212) look the same except the second one is ignoring the returned pointers as it is just making sure things are consistent.

It feels like there should be a way of turning the loops inside out and making this a method of ComputeDefUse and much cleaner overall, but I haven't figured out what that would be.  I'll think on it some more.

Reply:
I ended up rewriting this to be simpler and not need these extra functions -- now all the tests for correct and matching def/use are in the evalLoop that needs to be called to evaluate each iteration.  This avoids the separate check for that, and actually allows unrolling more cases (where the increment is in the loop body instead of in the loop increment)